pkg/mcp/creds/pollingWatcher.go

```diff
@@ -0,0 +1,202 @@
+// The supplied stop channel can be used to stop the go-routine and the watch.
+func PollFiles(stopCh <-chan struct{}, credentials *Options) (CertificateWatcher, error) {
+	// TODO: Make interval configurable
+	return pollFiles(stopCh, credentials, time.Minute)
+}
+
+func pollFiles(stopCh <-chan struct{}, credentials *Options, interval time.Duration) (CertificateWatcher, error) {
+	w := &pollingWatcher{
+		options:      *credentials,
+		pollInterval: interval,
+		stopCh:       stopCh,
+	}
+
+	if err := w.start(); err != nil {
+		return nil, err
+	}
+
+	return w, nil
+}
+
+// start watching and stop when the stopCh is closed. Returns an error if the initial load of the certificate
+// fails.
+func (p *pollingWatcher) start() error {
+	// Load CA Cert file
+	caCertPool, err := loadCACert(p.options.CACertificateFile)
+	if err != nil {
+		return err
+	}
+
+	if err = p.loadFiles(); err != nil {
+		return err
+	}
+
+	scope.Debugf("Begin polling certificate files: %s, %s: ",
+		p.options.CertificateFile, p.options.KeyFile)
+
+	go p.poll()
+
+	p.caCertPool = caCertPool
+
+	return nil
+}
+
+func (p *pollingWatcher) poll() {
+	t := time.NewTicker(p.pollInterval)
+	for {
+		select {
+		case <-t.C:
+			err := p.loadFiles()
+			if err != nil {
+				scope.Errorf("Error polling certificate files: %v", err)
+			}
+
+		case <-p.stopCh:
+			t.Stop()
+			scope.Debug("stopping poll of certificate file changes")
+			return
+		}
+	}
+}
+
+func (p *pollingWatcher) loadFiles() (err error) {
+	p.certMutex.Lock()
+	defer p.certMutex.Unlock()
+
+	defer func() {
+		p.pollErr = err
+	}()
+
+	var newKeyHash, newCertHash string
+
+	// Go through files and stat.
+	if newKeyHash, err = getHashSum(p.options.KeyFile); err != nil {
+		err = fmt.Errorf("unable to read key file(%q): %v", p.options.KeyFile, err)
+		return
+	}
+
+	if newCertHash, err = getHashSum(p.options.CertificateFile); err != nil {
+		err = fmt.Errorf("unable to read cert file(%q): %v", p.options.CertificateFile, err)
+		return
+	}
+
+	if newKeyHash != p.keyHash || newCertHash != p.certHash {
+		var cert tls.Certificate
+		cert, err = loadCertPair(p.options.CertificateFile, p.options.KeyFile)
+		if err != nil {
+			err = fmt.Errorf("unable load cert files as pair: %v", err)
+			return
+		}
+
+		p.cert = cert
+		p.keyHash = newKeyHash
+		p.certHash = newCertHash
+	}
+
+	return
+}
+
+func (p *pollingWatcher) pollError() error {
+	p.certMutex.Lock()
+	defer p.certMutex.Unlock()
+	return p.pollErr
+}
+
+// Get the currently loaded certificate.
+func (p *pollingWatcher) Get() tls.Certificate {
+	p.certMutex.Lock()
+	defer p.certMutex.Unlock()
+	return p.cert
+}
+
+// getHashSum is a helper func to calculate sha1 sum.
+func getHashSum(file string) (string, error) {
+	f, err := os.Open(file)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	r := bufio.NewReader(f)
+
+	h := sha1.New()
+
+	_, err = io.Copy(h, r)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%x", h.Sum(nil)), nil
```

Comment:
Would it be sufficient to compare the byte hash directly to avoid the extra byte-to-string conversion cost?

Reply:
Done.